In [51]:
# Install required packages
# !pip install python-dotenv
# !pip install numpy
# !pip install pandas
# !pip install matplotlib
# !pip install google-api-python-client
# !pip install mysql-connector-python
# !pip install sqlalchemy
# !pip install vaderSentiment
# !pip install gradio_client
# !pip install wordcloud

In [1]:
# Import required packages
from dotenv import load_dotenv
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from googleapiclient.discovery import build
import mysql.connector
from sqlalchemy import create_engine
from datetime import datetime
import re
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from gradio_client import Client
from wordcloud import WordCloud
from tqdm import tqdm

C:\Users\Jens\DataScience\Projects\youtube-channel-analytics\venv_development\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
# Load environment variables from .env file
load_dotenv()

# Get YouTube API key from .env 
youtube_api_key = os.getenv("youtube_api_key")

# Get local MySQL server username from .env
mysql_user = os.getenv("mysql_user")
# Get local MySQL server password from .env
mysql_password = os.getenv("mysql_password")

# Get AWS MySQL server username from .env
aws_mysql_user = os.getenv("aws_mysql_user")
# Get AWS MySQL server password from .env
aws_mysql_password = os.getenv("aws_mysql_password")

# Get Hugging Face Access Token from .env
huggingface_access_token = os.getenv("huggingface_access_token")

# Extract

In [4]:
# Build the YouTube service object
youtube = build("youtube", "v3", developerKey=youtube_api_key)

## Channel data

In [8]:
# Select channels
channel_names = ["AlexTheAnalyst", "LukeBarousse", "Thuvu5"]

# Initialize an empty list to store dictionaries for each channel
channels_ls = []

# Initialize an empty list to store uploads playlist IDs of all channels
uploads_playlist_ids = []

# Loop through each channel
for channel_name in channel_names:
    # Get channel data using the YouTube Channels API
    # Note: Uses 1 out of 10.000 units from the daily usage limit 
    channel_data = youtube.channels().list(part="statistics,snippet,contentDetails", forHandle=channel_name).execute()  

    # Extract channel data in dictionary format
    channel_dict = {
        "channel_id": channel_data["items"][0]["id"],
        "channel_name": channel_data["items"][0]["snippet"]["title"],
        "views": int(channel_data["items"][0]["statistics"]["viewCount"]),
        "videos": int(channel_data["items"][0]["statistics"]["videoCount"]),
        "subscribers": int(channel_data["items"][0]["statistics"]["subscriberCount"])
    }
    
    try:
        # Try to get channel thumbnail in maximum resolution
        channel_dict["thumbnail_url"] = channel_data["items"][0]["snippet"]["thumbnails"]["maxres"]["url"]
    except KeyError:
        try:
            # If maxres is not available, get high resolution
            channel_dict["thumbnail_url"] = channel_data["items"][0]["snippet"]["thumbnails"]["high"]["url"]
        except KeyError:
            # If high resolution is not available, get default resolution
            channel_dict["thumbnail_url"] = channel_data["items"][0]["snippet"]["thumbnails"]["default"]["url"]
    
    # Append channel data in dictionary format to the list
    channels_ls.append(channel_dict)
    
    # Append uploads playlist ID to the list 
    uploads_playlist_ids.append(channel_data["items"][0]["contentDetails"]["relatedPlaylists"]["uploads"])

# Convert list of dictionaries to pandas DataFrame
channel_df = pd.DataFrame(channels_ls) 
channel_df

channel_id           channel_name     views  videos  \
0  UC7cs8q-gJRlGwj4A8OmCmXg       Alex The Analyst  36999411     312   
1  UCLLw7jmFsvfIVaUFsLs8mlQ          Luke Barousse  22581819     158   
2  UCJQJAI7IjbLcpsjWdSzYz0Q  Thu Vu data analytics   8276248      88   

   subscribers                                      thumbnail_url  
0       822000  https://yt3.ggpht.com/ytc/AIdro_l9wLnClpLKJeVm...  
1       454000  https://yt3.ggpht.com/ytc/AIdro_my6YXWfudW8qM_...  
2       244000  https://yt3.ggpht.com/s3HLl-uzqEaqww2tkWKgjLFf...

## Video data

In [9]:
# Initialize an empty list to store dictionaries for each video
videos_ls = []

# Loop through each channel's uploads playlist
for uploads_playlist_id in uploads_playlist_ids:
    # Initialize next_page_token to None
    next_page_token = None

    # Loop through each video in the playlist
    while True:
        # Get playlist data using the YouTube PlaylistItems API 
        # Note: Each loop uses 1 out of 10.000 units from the daily usage limit (1 unit for 50 videos)
        playlist_data = youtube.playlistItems().list(
            part="snippet", 
            playlistId=uploads_playlist_id, 
            maxResults=50,
            pageToken=next_page_token
        ).execute()

        # Initialize an empty list to store video IDs
        video_ids = []

        # Extract video IDs from the playlist data
        video_ids += [video_data["snippet"]["resourceId"]["videoId"] for video_data in playlist_data["items"]]

        # Get video data using the YouTube Videos API 
        # Note: Uses 1 out of 10.000 units from the daily usage limit (1 unit for 50 videos)
        video_data = youtube.videos().list(part="statistics,snippet,contentDetails", id=video_ids).execute()    

        # Loop through each video 
        for video in video_data["items"]:
            # Extract video data in dictionary format
            video_dict = {
                "video_id": video["id"],
                "channel_id": video["snippet"]["channelId"],
                "video_title": video["snippet"]["title"],
                "video_description": video["snippet"]["description"],
                "published_at": datetime.strptime(video["snippet"]["publishedAt"], "%Y-%m-%dT%H:%M:%SZ"),
                "video_duration": video["contentDetails"]["duration"],
                "views": int(video["statistics"]["viewCount"]),
                "likes": int(video["statistics"]["likeCount"]),
                "comments": int(video["statistics"]["commentCount"])
            }

            try:
                # Try to get thumbnail in maximum resolution
                video_dict["thumbnail_url"] = video["snippet"]["thumbnails"]["maxres"]["url"]
            except KeyError:
                try:
                    # If maxres is not available, get high resolution
                    video_dict["thumbnail_url"] = video["snippet"]["thumbnails"]["high"]["url"]
                except KeyError:
                    # If high resolution is not available, get default resolution
                    video_dict["thumbnail_url"] = video["snippet"]["thumbnails"]["default"]["url"]

            # Append video data in dictionary format to the list
            videos_ls.append(video_dict)

        # Get the next page token
        next_page_token = playlist_data.get("nextPageToken")

        # Exit the loop if there are no more pages
        if next_page_token is None:
            break
        
# Convert list of dictionaries to pandas DataFrame
videos_df = pd.DataFrame(videos_ls)    
videos_df

video_id                channel_id  \
0    ZYps6TmBkWk  UC7cs8q-gJRlGwj4A8OmCmXg   
1    zv1nfZTYpio  UC7cs8q-gJRlGwj4A8OmCmXg   
2    rn0dSsYXhIE  UC7cs8q-gJRlGwj4A8OmCmXg   
3    siHou7lObbo  UC7cs8q-gJRlGwj4A8OmCmXg   
4    wJiJXD7MVoY  UC7cs8q-gJRlGwj4A8OmCmXg   
..           ...                       ...   
554  5LWoJAh-kww  UCJQJAI7IjbLcpsjWdSzYz0Q   
555  _RzoHVWKwq4  UCJQJAI7IjbLcpsjWdSzYz0Q   
556  hWKLO7GtpiU  UCJQJAI7IjbLcpsjWdSzYz0Q   
557  dBZqggW22rs  UCJQJAI7IjbLcpsjWdSzYz0Q   
558  QDdqsFCIxIk  UCJQJAI7IjbLcpsjWdSzYz0Q   

                                           video_title  \
0    Azure Account Setup + $200 Free Credits | Azur...   
1    Building a Fully Interactive Web App using Shi...   
2    Shiny for Python Components and Layouts | Cust...   
3    Shiny for Python Setup and Install | Deploying...   
4    Is the Economy on the Brink of a Recession? | ...   
..                                                 ...   
554  Data Analyst Skill Stack // How I Became A Dat...   
555  Effective visual note-taking on iPad + Downloa...   
556  Data scientist is NOT the only SEXY job // All...   
557  How I take notes on iPad Pro | Notion, Notes, ...   
558  How to stay focused and productive online with...   

                                     video_description        published_at  \
0    In this lesson we will be setting up and walki... 2024-07-16 12:00:38   
1    In this video we are building a Full Shiny App... 2024-07-09 12:01:07   
2    In this video we look at how we can customize ... 2024-07-02 12:01:00   
3    In this video we are going to install Shiny an... 2024-06-25 12:00:53   
4    The Economy isn't looking good and there's a l... 2024-06-18 12:00:06   
..                                                 ...                 ...   
554  👩🏻‍💻 My laptop and iPad for doing DS/ study 👉 ... 2021-06-16 10:16:47   
555  Hi there! In this video I’m showing you how to... 2021-06-09 06:58:21   
556  Hi there! In this video I explain different da... 2021-06-01 22:51:42   
557  My note-taking system tour in Notion and tips ... 2021-05-16 22:47:12   
558  Hello Youtube! This is my first ever Youtube v... 2021-04-30 16:03:53   

    video_duration  views  likes  comments  \
0          PT6M37S   4056    201        25   
1         PT59M52S   6106    235        20   
2          PT24M8S   3508    114         7   
3         PT14M24S   6138    222        31   
4         PT21M19S  11367    693       124   
..             ...    ...    ...       ...   
554       PT10M35S  33230   1793        70   
555        PT8M19S  12599    420        15   
556        PT14M4S   9041    295        20   
557         PT9M7S  12001    281        21   
558        PT4M34S   5098    176        14   

                                         thumbnail_url  
0    https://i.ytimg.com/vi/ZYps6TmBkWk/maxresdefau...  
1    https://i.ytimg.com/vi/zv1nfZTYpio/maxresdefau...  
2    https://i.ytimg.com/vi/rn0dSsYXhIE/maxresdefau...  
3    https://i.ytimg.com/vi/siHou7lObbo/maxresdefau...  
4    https://i.ytimg.com/vi/wJiJXD7MVoY/maxresdefau...  
..                                                 ...  
554  https://i.ytimg.com/vi/5LWoJAh-kww/maxresdefau...  
555  https://i.ytimg.com/vi/_RzoHVWKwq4/maxresdefau...  
556  https://i.ytimg.com/vi/hWKLO7GtpiU/maxresdefau...  
557  https://i.ytimg.com/vi/dBZqggW22rs/maxresdefau...  
558  https://i.ytimg.com/vi/QDdqsFCIxIk/maxresdefau...  

[559 rows x 10 columns]

## Comments data

In [10]:
# Initialize an empty list to store comments
comments_ls = []

# Loop through each video
for video_id in videos_df["video_id"].values:
    # Initialize next_page_token to None
    next_page_token = None

    # Loop through data batches of 100 comments 
    while True:
        try:
            # Get data from 100 comments using the YouTube CommentThreads API 
            # Note: Each loop uses 1 out of 10.000 units from the daily usage limit (1 unit for 100 comments)
            comments_data = youtube.commentThreads().list(
                part="snippet", 
                videoId=video_id, 
                maxResults=100,
                pageToken=next_page_token
            ).execute()
        # Handle error if e.g. video comments are disabled
        except Exception as e:
            print(f"Failed to get comments for video {video_id}.")

        # Loop through each comment
        for comment in comments_data["items"]:
            # Extract comment data in dictionary format
            comment_dict = {
                "comment_id": comment["snippet"]["topLevelComment"]["id"],
                "video_id": comment["snippet"]["topLevelComment"]["snippet"]["videoId"],
                "channel_id": comment["snippet"]["topLevelComment"]["snippet"]["channelId"],
                "comment_text": comment["snippet"]["topLevelComment"]["snippet"]["textOriginal"],
                "published_at": datetime.strptime(comment["snippet"]["topLevelComment"]["snippet"]["publishedAt"], "%Y-%m-%dT%H:%M:%SZ")
            }
            # Append comment data dictionary to the list
            comments_ls.append(comment_dict)

        # Get the next page token
        next_page_token = comments_data.get("nextPageToken")

        # Exit the loop if there are no more pages
        if next_page_token is None: 
            break
        
# Convert list of dictionaries to pandas DataFrame
comments_df = pd.DataFrame(comments_ls)    
comments_df 

Failed to get comments for video W_s4c1vLGXM.
Failed to get comments for video qfyynHBFOsM.
Failed to get comments for video G4syHs3M82E.
Failed to get comments for video LJtFgnHGAos.
Failed to get comments for video r9imv1z82jQ.


comment_id     video_id                channel_id  \
0      UgzSfHTFO8ZWcg_maUF4AaABAg  ZYps6TmBkWk  UC7cs8q-gJRlGwj4A8OmCmXg   
1      Ugym_vQM6LOs4x1mV5N4AaABAg  ZYps6TmBkWk  UC7cs8q-gJRlGwj4A8OmCmXg   
2      UgwuWMq45Mw5DmWuCkJ4AaABAg  ZYps6TmBkWk  UC7cs8q-gJRlGwj4A8OmCmXg   
3      Ugz9XMUa83zfOqZQ17h4AaABAg  ZYps6TmBkWk  UC7cs8q-gJRlGwj4A8OmCmXg   
4      Ugx5zQmHZ_vs6PPI7Ht4AaABAg  ZYps6TmBkWk  UC7cs8q-gJRlGwj4A8OmCmXg   
...                           ...          ...                       ...   
58222  UgykL4IfQ7CLbONTW5t4AaABAg  QDdqsFCIxIk  UCJQJAI7IjbLcpsjWdSzYz0Q   
58223  UgzyWHk3Kx6N3sbElMJ4AaABAg  QDdqsFCIxIk  UCJQJAI7IjbLcpsjWdSzYz0Q   
58224  Ugw1X4GuVl6jdAB8sFJ4AaABAg  QDdqsFCIxIk  UCJQJAI7IjbLcpsjWdSzYz0Q   
58225  UgzDqSm_tOqxIbgclvF4AaABAg  QDdqsFCIxIk  UCJQJAI7IjbLcpsjWdSzYz0Q   
58226  Ugy9O3Xlo1_2Oqb8G-R4AaABAg  QDdqsFCIxIk  UCJQJAI7IjbLcpsjWdSzYz0Q   

                                            comment_text        published_at  
0                                              i'm ready 2024-07-22 23:04:43  
1                                         Please help me 2024-07-21 16:04:03  
2      Hey Alex the social  links in your channel des... 2024-07-19 23:44:53  
3      The future of Binance: an exclusive interview ... 2024-07-19 06:50:19  
4      Thanks Alex! Great series, much needed! Could ... 2024-07-18 23:40:08  
...                                                  ...                 ...  
58222                thank you, please i need you help!! 2022-03-07 11:10:44  
58223  Very diff but great inputs from similar topic ... 2022-02-20 00:42:23  
58224  Replying email and control backspace was new f... 2021-08-02 15:15:34  
58225                                  Valuable insight😇 2021-07-14 10:17:53  
58226                 good tips for working time at home 2021-05-04 09:26:36  

[58227 rows x 5 columns]

# Transform

## Convert video duration

In [11]:
# Function to convert the YouTube video duration from ISO 8601 format (str) to seconds (int)
def convert_iso8601_duration(duration):
    # Regular expression to match hours, minutes, and seconds
    time_extractor = re.compile(r'PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?')
    # Extract hours, minutes, and seconds
    extracted = time_extractor.match(duration)
    if extracted:
        hours = int(extracted.group(1)) if extracted.group(1) else 0
        minutes = int(extracted.group(2)) if extracted.group(2) else 0
        seconds = int(extracted.group(3)) if extracted.group(3) else 0
        # Return total seconds
        total_seconds = hours * 3600 + minutes * 60 + seconds
        return total_seconds
    else:
        return 0

In [12]:
# Convert video duration in pandas DataFrame
videos_df["video_duration"] = videos_df["video_duration"].apply(convert_iso8601_duration)

## Sentiment analysis

To identify the most effective sentiment analysis method for YouTube comments, we will compare three models: VADER, DistilBERT, and RoBERTa. Each will be applied to a dataset of 2000 randomly selected comments. VADER employs a rule-based approach, while DistilBERT and RoBERTa are machine learning-based approaches.

### Fetch example comments (optional) 
For testing, skip the data extraction from the YouTube API and instead fetch 2000 random comments from the comments table of the AWS RDS MySQL database.

Note: Make sure to establish an SSH tunnel via PuTTY to connect to the AWS RDS MySQL server instance through the EC2 instance.

In [24]:
try:
    # Create an SQLAlchemy engine for interacting with the MySQL database
    engine = create_engine(f"mysql+mysqlconnector://{aws_mysql_user}:{aws_mysql_password}@localhost:3308/youtube_analytics") 
    
    # MySQL query to fetch 2000 random comments
    query = "SELECT * FROM comments ORDER BY RAND() LIMIT 2000"  
    
    # Execute query to load the comments from the MySQL comments table into a pandas DataFrame
    try:
        random_comments_df = pd.read_sql(query, engine)
        print("2000 random comments successfully fetched from the AWS MySQL database.")
    except Exception as e:
        print("Error fetching random comments from the AWS MySQL database:", e)
    
except Exception as e:
    # Print error if exception occurs when connecting to the database 
    print("Error connecting to AWS MySQL database:", e)

finally:
    # Close the database connection
    engine.dispose()

2000 random comments successfully fetched from the AWS MySQL database.


In [26]:
# Drop comments with missing comment ID or comment text (e.g. because the comment was deleted, marked as spam, or private)
random_comments_df = random_comments_df.dropna(subset=["comment_id", "comment_text"])

In [28]:
# Drop duplicate comments
random_comments_df = random_comments_df.drop_duplicates()

In [30]:
df = random_comments_df.copy()

In [32]:
# Keep only the columns for comment ID and comment text 
random_comments_df = random_comments_df[["comment_id", "comment_text"]].copy()
random_comments_df.head()

comment_id  \
0  Ugzz8xiGHMAYlk0qY-14AaABAg   
1  Ugx8NTehfQyBD5g6Sl14AaABAg   
2  UgxOFAXWoPBs25sSe8t4AaABAg   
3  UgxdcYZEs7wm5lGmIft4AaABAg   
4  UgyavIAxjby0vYmb7sd4AaABAg   

                                        comment_text  
0  Are the majority of the 120 companies listed a...  
1                                 Awesome video bruh  
2  Sql databases are going to be taken over by Fa...  
3  I have been trying to mess around with the ext...  
4  Thanks for this Alex, I have started business ...

### VADER
VADER (Valence Aware Dictionary for sEntiment Reasoning) is a sentiment analysis tool designed specifically for social media texts. It employs a rule-based approach, leveraging a pre-defined lexicon of words and phrases along with grammatical and syntactical rules to determine the sentiment of a given text. Based on the sentiment scores, texts can be classified into positive, negative or neutral.

In [13]:
# Initialize the VADER sentiment intensity analyzer
vader_sia = SentimentIntensityAnalyzer()

# Define a function to get sentiment scores
def get_sentiment(text):
    return vader_sia.polarity_scores(text)["compound"]

# Define a function to categorize the sentiment scores into positve, negative, or neutral
def categorize_sentiment(score):
    if score > 0.05:
        return "positive"
    elif score < -0.05:
        return "negative"
    else:
        return "neutral"

# Apply sentiment analysis to each comment and store sentiment scores in new column 
random_comments_df["vader_score"] = random_comments_df["comment_text"].apply(get_sentiment)

# Apply sentiment categorization to each comment and store sentiment category in new column 
random_comments_df["vader_sentiment"] = random_comments_df["vader_score"].apply(categorize_sentiment)

In [15]:
# Display the first few rows to verify the results
pd.set_option("display.max_colwidth", None)
random_comments_df[["comment_text", "vader_score", "vader_sentiment"]].head()

comment_text  \
0                                                                                                                                                                                         Thank you very much <3   
1  In the last query, you used a subquery in the WHERE clause to filter for employees over 30. Could you also have accomplished that by joining the two tables and adding a WHERE EmployeeDemographics.Age > 30?   
2                                                                                        Thanks so much for the informative video! Though 10 hours per day for 6 months may stretch my time budget a bit ;) 1:51   
3        Thank you Luke! Your videos are extremely helpful and practical.\n\nOne small question, I noticed that you used the IIF function instead of IF . Does that make any difference in that particular case?   
4                                                                                                                                                                                    Really, really informative!   

   vader_score vader_sentiment  
0       0.6887        positive  
1       0.4404        positive  
2       0.6239        positive  
3       0.7082        positive  
4       0.0000         neutral

### DistilBERT
DistilBERT (Distilled BERT) is a machine learning-based model that is a smaller, faster, and more efficient version of BERT (Bidirectional Encoder Representations from Transformers). With 66 million parameters, the [distilbert-base-uncased-finetuned-sst-2-english](https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english) model by Hugging Face maintains around 97% of BERT's performance while being significantly faster and lighter. Fine-tuned on the Stanford Sentiment Treebank (SST-2) dataset, it achieves 91.3% accuracy in sentiment classification, close to BERT's 92.7%. The model performs binary sentiment classification, identifying text as either positive or negative.

In [68]:
# Initialize Gradio client to connect to a Gradio server running locally
# Note: Ensure the Gradio web application is running locally before making a request
client = Client("http://127.0.0.1:7860/")

# Function to get sentiment scores of comments in batches via API requests
def get_sentiment(df, model_choice, batch_size=500):  
    # Initialize a dictionary to store sentiment analysis results
    result = {
        f"{model_choice.lower()}_sentiment": [],
        f"{model_choice.lower()}_confidence": []
    }

    # Iterate over the DataFrame in batches
    for i in range(0, len(df), batch_size):
        # Extract the current batch
        batch_df = df.iloc[i:i+batch_size]    
        
        # Prepare data of current batch in JSON format for API request 
        batch_data = {
            "comment_text": batch_df["comment_text"].tolist(),
            "model_choice": model_choice
        }
        
        # Send batch to API and get results
        batch_result = client.predict(
            batch_data,
            api_name="/predict"
        )

        # Append the results of the current batch to the corresponding lists in the result dictionary
        result[f"{model_choice.lower()}_sentiment"].extend(batch_result[f"{model_choice.lower()}_sentiment"])
        result[f"{model_choice.lower()}_confidence"].extend(batch_result[f"{model_choice.lower()}_confidence"])

    # Return the dictionary with the sentiment analysis results of all comments
    return result

Loaded as API: http://127.0.0.1:7860/ ✔


In [21]:
# Apply DistilBERT sentiment analysis
results_json = get_sentiment(random_comments_df, model_choice="DistilBERT")

# Convert results from JSON to Pandas DataFrame
results_df = pd.DataFrame(results_json)

# Add DistilBERT sentiment and confidence as new columns to the original DataFrame
random_comments_df["distilbert_sentiment"] = results_df["distilbert_sentiment"]
random_comments_df["distilbert_confidence"] = results_df["distilbert_confidence"]

Note: Took 2 min 26 sec to run on local machine for 2000 comments in batches of 500 per API request.

In [23]:
# Display the first few rows to verify the results
random_comments_df.head()

comment_text  \
0                                                                                                                                                                                         Thank you very much <3   
1  In the last query, you used a subquery in the WHERE clause to filter for employees over 30. Could you also have accomplished that by joining the two tables and adding a WHERE EmployeeDemographics.Age > 30?   
2                                                                                        Thanks so much for the informative video! Though 10 hours per day for 6 months may stretch my time budget a bit ;) 1:51   
3        Thank you Luke! Your videos are extremely helpful and practical.\n\nOne small question, I noticed that you used the IIF function instead of IF . Does that make any difference in that particular case?   
4                                                                                                                                                                                    Really, really informative!   

   vader_score vader_sentiment distilbert_sentiment  distilbert_confidence  
0       0.6887        positive             positive                  1.000  
1       0.4404        positive             negative                  0.998  
2       0.6239        positive             positive                  0.990  
3       0.7082        positive             positive                  0.999  
4       0.0000         neutral             positive                  1.000

### RoBERTa 
RoBERTa (Robustly Optimized BERT Pretraining Approach) is an enhanced version of BERT, featuring 125 million parameters and improved performance due to larger datasets and optimized hyperparameters. The [twitter-roberta-base-sentiment-latest](https://huggingface.co/cardiffnlp/twitter-roberta-base-sentiment-latest) model by [cardiffnlp](https://huggingface.co/cardiffnlp) is trained on approximately 124 million tweets and fine-tuned using the TweetEval benchmark. Unlike DistilBERT, RoBERTa performs ternary classification, categorizing text as positive, negative, or neutral. 

In [77]:
# Apply RoBERTa sentiment analysis
results_json = get_sentiment(random_comments_df, model_choice="RoBERTa")

# Convert results from JSON to Pandas DataFrame
results_df = pd.DataFrame(results_json)

# Add RoBERTa sentiment and confidence as new columns to the original DataFrame
random_comments_df["roberta_sentiment"] = results_df["roberta_sentiment"]
random_comments_df["roberta_confidence"] = results_df["roberta_confidence"]

Note: Took 4 min 56 sec to run on local machine for 2000 comments in batches of 500 per API request.

In [73]:
# Display the first few rows to verify the results
random_comments_df.head()

comment_text  \
0                                                                                                                                                                                         Thank you very much <3   
1  In the last query, you used a subquery in the WHERE clause to filter for employees over 30. Could you also have accomplished that by joining the two tables and adding a WHERE EmployeeDemographics.Age > 30?   
2                                                                                        Thanks so much for the informative video! Though 10 hours per day for 6 months may stretch my time budget a bit ;) 1:51   
3        Thank you Luke! Your videos are extremely helpful and practical.\n\nOne small question, I noticed that you used the IIF function instead of IF . Does that make any difference in that particular case?   
4                                                                                                                                                                                    Really, really informative!   

   vader_score vader_sentiment distilbert_sentiment  distilbert_confidence  \
0       0.6887        positive             positive                  1.000   
1       0.4404        positive             negative                  0.998   
2       0.6239        positive             positive                  0.990   
3       0.7082        positive             positive                  0.999   
4       0.0000         neutral             positive                  1.000   

  roberta_sentiment  roberta_confidence  
0          positive               0.960  
1           neutral               0.897  
2          positive               0.956  
3          positive               0.906  
4          positive               0.974

In [124]:
# Save 500 comments with sentiment analysis results from VADER, DistilBERT, and RoBERTa as CSV file
random_comments_df.head(500).to_csv("sentiment_analysis_method_comparison.csv", index=False)

Comparing the sentiment analysis results, RoBERTa showed the best performance.

### RoBERTa on Hugging Face Spaces

In [34]:
# Initialize Gradio client to connect to the RoBERTa Sentiment Analysis API on Hugging Face Spaces
# Note: Ensure the Gradio web application with RoBERTa Sentiment Analysis API is running on Hugging Face Spaces
client = Client("JensBender/roberta-sentiment-analysis-api", hf_token=huggingface_access_token)  

# Function to get sentiment scores of comments in batches via API requests
def get_sentiment(df, batch_size=500):  
    # Initialize a dictionary to store sentiment analysis results
    result = {
        "comment_id": [],
        "sentiment": [],
        "sentiment_confidence": []
    }

    # Calculate the number of batches
    num_batches = (len(df) + batch_size - 1) // batch_size
    
    # Iterate over the DataFrame in batches
    for i in tqdm(range(0, len(df), batch_size), total=num_batches, desc="Processing Batches"):
        # Extract the current batch
        batch_df = df.iloc[i:i+batch_size]    
        
        # Prepare data of current batch in JSON format for API request 
        batch_data = {
            "comment_id": batch_df["comment_id"].tolist(),
            "comment_text": batch_df["comment_text"].tolist()
        }
        
        # Send batch to API and get results
        batch_result = client.predict(
            batch_data,
            api_name="/predict"
        )
           
        # Append the results of the current batch to the corresponding lists in the result dictionary
        result["comment_id"].extend(batch_result["comment_id"])
        result["sentiment"].extend(batch_result["roberta_sentiment"])
        result["sentiment_confidence"].extend(batch_result["roberta_confidence"])

    # Return the dictionary with the sentiment analysis results of all comments
    return result

Loaded as API: https://jensbender-roberta-sentiment-analysis-api.hf.space ✔


In [36]:
# Apply RoBERTa sentiment analysis
results_json = get_sentiment(random_comments_df)

# Convert results from JSON to Pandas DataFrame
results_df = pd.DataFrame(results_json)

Processing Batches: 100%|███████████████████████████████████████████████████████████████| 4/4 [21:26<00:00, 321.69s/it]


In [38]:
# Merge the RoBERTa sentiment and confidence columns with the original DataFrame
random_comments_df = pd.merge(df, results_df, on="comment_id")

In [40]:
# Display the first few rows to verify the results
random_comments_df.head()

comment_id     video_id                channel_id  \
0  Ugzz8xiGHMAYlk0qY-14AaABAg  9RRQtNnq3s0  UC7cs8q-gJRlGwj4A8OmCmXg   
1  Ugx8NTehfQyBD5g6Sl14AaABAg  qzZU6LAtIig  UC7cs8q-gJRlGwj4A8OmCmXg   
2  UgxOFAXWoPBs25sSe8t4AaABAg  9ur0OpMADuM  UC7cs8q-gJRlGwj4A8OmCmXg   
3  UgxdcYZEs7wm5lGmIft4AaABAg  QILNlRvJlfQ  UC7cs8q-gJRlGwj4A8OmCmXg   
4  UgyavIAxjby0vYmb7sd4AaABAg  rGx1QNdYzvs  UC7cs8q-gJRlGwj4A8OmCmXg   

                                        comment_text         published_at  \
0  Are the majority of the 120 companies listed a...  2021-04-14 11:53:37   
1                                 Awesome video bruh  2020-03-11 15:15:31   
2  Sql databases are going to be taken over by Fa...  2024-07-30 18:46:36   
3  I have been trying to mess around with the ext...  2021-05-23 22:46:25   
4  Thanks for this Alex, I have started business ...  2023-02-03 23:06:26   

  sentiment  sentiment_confidence  
0   neutral                 0.866  
1  positive                 0.982  
2   neutral                 0.823  
3   neutral                 0.685  
4  positive                 0.853

# Load

## Into local MySQL database

In [ ]:
# Connect to local MySQL database
connection = mysql.connector.connect(
    host = "localhost",
    port = 3306,
    user = mysql_user,
    password = mysql_password,
    database = "youtube_analytics"
)

# Create a cursor object to execute SQL queries
cursor = connection.cursor()

# Drop existing MySQL tables 
tables_to_drop = ["comments", "videos", "channels"]
for table in tables_to_drop:
    cursor.execute(f"DROP TABLE IF EXISTS {table};")
        
try:
    # Create an SQLAlchemy engine for interacting with the MySQL database
    engine = create_engine(f"mysql+mysqlconnector://{mysql_user}:{mysql_password}@localhost:3306/youtube_analytics") 
    
    # Load the YouTube channels DataFrame into the MySQL channels table
    try:
        channel_df.to_sql("channels", con=engine, if_exists="replace", index=False)
        print("Channels data successfully loaded into local MySQL database.")
    except Exception as e:
        print("Error loading channels data:", e)
    
    # Load the YouTube videos DataFrame into the MySQL videos table
    try:
        videos_df.to_sql("videos", con=engine, if_exists="replace", index=False)
        print("Videos data successfully loaded into local MySQL database.")
    except Exception as e:
        print("Error loading videos data:", e)
    
    # Load the YouTube comments DataFrame into the MySQL comments table
    try:
        comments_df.to_sql("comments", con=engine, if_exists="replace", index=False)
        print("Comments data successfully loaded into local MySQL database.")
    except Exception as e:
        print("Error loading comments data:", e)
    
except Exception as e:
    # Print error if exception occurs when connecting to the database 
    print("Error connecting to local MySQL database:", e)

finally:
    # Close the cursor and connection to free up resources
    cursor.close()
    connection.close()

## Into AWS MySQL database
Note: Make sure to establish an SSH tunnel via PuTTY to connect to the AWS RDS MySQL server instance through the EC2 instance.

In [ ]:
# Connect to AWS MySQL database
connection = mysql.connector.connect(
    host = "localhost",
    port = 3308,
    user = aws_mysql_user,
    password = aws_mysql_password,
    database = "youtube_analytics"
)

# Create a cursor object to execute SQL queries
cursor = connection.cursor()

# Drop existing MySQL tables 
tables_to_drop = ["comments", "videos", "channels"]
for table in tables_to_drop:
    cursor.execute(f"DROP TABLE IF EXISTS {table};")
        
try:
    # Create an SQLAlchemy engine for interacting with the MySQL database
    engine = create_engine(f"mysql+mysqlconnector://{aws_mysql_user}:{aws_mysql_password}@localhost:3308/youtube_analytics") 
    
    # Load the YouTube channels DataFrame into the MySQL channels table
    try:
        channel_df.to_sql("channels", con=engine, if_exists="replace", index=False)
        print("Channels data successfully loaded into AWS MySQL database.")
    except Exception as e:
        print("Error loading channels data:", e)
    
    # Load the YouTube videos DataFrame into the MySQL videos table
    try:
        videos_df.to_sql("videos", con=engine, if_exists="replace", index=False)
        print("Videos data successfully loaded into AWS MySQL database.")
    except Exception as e:
        print("Error loading videos data:", e)
    
    # Load the YouTube comments DataFrame into the MySQL comments table
    try:
        comments_df.to_sql("comments", con=engine, if_exists="replace", index=False)
        print("Comments data successfully loaded into AWS MySQL database.")
    except Exception as e:
        print("Error loading comments data:", e)
    
except Exception as e:
    # Print error if exception occurs when connecting to the database 
    print("Error connecting to AWS MySQL database:", e)

finally:
    # Close the cursor and connection to free up resources
    cursor.close()
    connection.close()

# Word clouds

Word clouds to visualize the main topics of each YouTube channel.

In [ ]:
# Loop through each channel
for channel_id in channel_df["channel_id"].values:
    # Print channel name
    print(channel_df[channel_df["channel_id"]==channel_id]["channel_name"].values[0])

    # Combine all video titles into a single string
    text = " ".join(videos_df[videos_df["channel_id"]==channel_id]["video_title"])
    
    # Create a WordCloud object
    wordcloud = WordCloud(width=800, height=400, background_color="white", random_state=7)

    # Create a word cloud of the video titles
    wordcloud.generate(text)

    # Display the word cloud 
    plt.figure(figsize=(10, 5))
    plt.imshow(wordcloud, interpolation="bilinear")
    plt.axis("off")  # Turn off the axis
    plt.show()